src/bootstrap/src/core/config/config.rs

```diff
@@ -28,6 +28,61 @@ use crate::utils::cache::{INTERNER, Interned};
 use crate::utils::channel::{self, GitInfo};
 use crate::utils::helpers::{self, exe, output, t};
 
+/// Each path in this list is considered "allowed" in the `download-rustc="if-unchanged"` logic.
+/// This means they can be modified and changes to these paths should never trigger a compiler build
+/// when "if-unchanged" is set.
+///
+/// NOTE: Paths must have the ":!" prefix to tell git to ignore changes in those paths during
+/// the diff check.
+///
+/// WARNING: Be cautious when adding paths to this list. If a path that influences the compiler build
+/// is added here, it will cause bootstrap to skip necessary rebuilds, which may lead to risky results.
+pub(crate) const RUSTC_IF_UNCHANGED_ALLOWED_PATHS: &[&str] = &[
+    ":!.github",
+    ":!.clang-format",
+    ":!.editorconfig",
+    ":!.git-blame-ignore-revs",
+    ":!.gitattributes",
+    ":!.gitignore",
+    ":!.gitmodules",
+    ":!.ignore",
+    ":!.mailmap",
+    ":!CODE_OF_CONDUCT.md",
+    ":!CONTRIBUTING.md",
+    ":!COPYRIGHT",
+    ":!INSTALL.md",
+    ":!LICENSE-APACHE",
+    ":!LICENSE-MIT",
+    ":!LICENSES",
+    ":!README.md",
+    ":!RELEASES.md",
+    ":!REUSE.toml",
+    ":!config.example.toml",
+    ":!configure",
+    ":!rust-bors.toml",
+    ":!rustfmt.toml",
+    ":!tests",
+    ":!triagebot.toml",
+    ":!x",
+    ":!x.ps1",
+    ":!x.py",
+    ":!src/bootstrap",
```

Comment:
This seems wrong? If this is changed, we probably *should* rebuild the compiler from scratch.

Similarly the CI setup (.github, src/ci, etc.) *for CI* should be "changes" that cause a compiler rebuild. Locally it seems reasonable for them to do that too, since they're unlikely to cause problems for local users but may still affect their tests (e.g., running CI Docker containers locally should rebuild the compiler not download it if CI scripts are changed).

Reply:
> This seems wrong? If this is changed, we probably should rebuild the compiler from scratch.

I thought about this too but realized it could risk merging a PR that would break the CI the next day(I mentioned this [here](https://github.com/rust-lang/rust/pull/131831#issuecomment-2424721864) in case you haven't seen it yet).

But this won't happen when we allow bootstrap changes in "if-unchanged" logic. If any bootstrap changes break the compiler build, we will catch them in the dist runners since we are building from scratch there.

Reply:
I'm confused by that argument - it seems true, but it seems like allowing bootstrap just postpones the problem to the next PR touching bootstrap (or anything else outside the allowlist). I'd assume that if we want to avoid that we need robust testing of some kind (e.g., always trying to use CI artifacts with some limited tests or something).

Reply:
I'm not sure what the best approach would be (it's something we should spend sometime on thinking), but this definitely needs improvement. That said, I think current state of this PR is already an improvement over the status quo. Bootstrap changes are already allowed in CI, and certain runners (like those not using ci-llvm) always use the in-tree compiler, so there's at least some coverage for "running compiler tests with a non-CI compiler while there are changes on bootstrap". Personally, I wouldn't block this PR as it's a lot better than what we currently have.

Reply:
Sure, done.

Reply:
We just had a confusing case of "test fails but it shouldn't" caused by a cached compiler being used despite a bootstrap change. So bootstrap changes definitely should lead to a rebuild.

Might be worth leaving a comment to that note, or else we'll likely have this same discussion again in 2 years.

Reply:
Is [this](https://github.com/rust-lang/rust/pull/131831/commits/a00fd7a965c455d137c887ebbb7ed580719f080c) enough?

Reply:
Yes, thanks :)